### PR TITLE
dependence analysis & legality check

### PR DESCRIPTION
**Pull request by nassimTchoulak**

### CODE REVIEW

File: src/tiramisu_core.cpp [hunk 1]
```diff
@@ -97,6 +97,26 @@ void codegen(const std::vector<tiramisu::buffer *> &arguments, const std::string
...
 }
 
+bool check_legality_of_function()
+{
+    function *fct = global::get_implicit_function();
+    return fct->check_legality_for_function() ;
+}
+
+
+void performe_full_dependecy_analysis()
```
Comment (rbaghdadi) [inline src/tiramisu_core.cpp:107]: Ok I see now the difference.

File: src/tiramisu_core.cpp [hunk 2]
```diff
@@ -4395,6 +4415,205 @@ void computation::skew(int L0 , int L1 , int f_i , int f_j)
...
+        return isl_stat_ok;
+    };
+    
+    isl_stat (*fun_ptr)(isl_basic_map * p,void * m) = (f) ;
+
+    isl_map_foreach_basic_map(my_map1,fun_ptr,(void * ) &all_basic_maps) ;
+
+    isl_map_foreach_basic_map(my_map2,fun_ptr,(void * ) &all_basic_maps) ;
+
+    isl_map_foreach_basic_map(my_map3,fun_ptr,(void * ) &all_basic_maps) ;
```
Comment (rbaghdadi) [inline src/tiramisu_core.cpp:4503]: No space before ";"

Space after ","

File: src/tiramisu_core.cpp [hunk 3]
```diff
@@ -4395,6 +4415,205 @@ void computation::skew(int L0 , int L1 , int f_i , int f_j)
...
+   assert(m1 == m2) ;
+
+    DEBUG(3, tiramisu::str_dump(" the current schedule of computation "+this->get_name()+" : "+std::string(isl_map_to_str(this->get_schedule()))));
+    DEBUG(3, tiramisu::str_dump(" the current schedule of computation "+second->get_name()+" : "+std::string(isl_map_to_str(second->get_schedule()))));
+
+    /* ==========================================
+       making schedules comparable by mapping to the same time space
+    */
+  
+    std::string empty ="" ;
```
Comment (rbaghdadi) [inline src/tiramisu_core.cpp:4523]: Space before ""

File: src/tiramisu_function.cpp [hunk 4]
```diff
@@ -2397,4 +2565,111 @@ const std::vector<std::string> tiramisu::function::get_invariant_names() const
...
+
+    this->calculate_dep_flow() ;
+    
+    DEBUG_INDENT(-4);
+
+}
+
+bool tiramisu::function::check_legality_for_function()
+{
+
```
Comment (rbaghdadi) [inline src/tiramisu_function.cpp:2587]: Drop space.

File: src/tiramisu_function.cpp [hunk 5]
```diff
@@ -2397,4 +2565,111 @@ const std::vector<std::string> tiramisu::function::get_invariant_names() const
...
+        left_comp = this->get_computation_by_name(left_computation_name)[0] ;
+        right_comp = this->get_computation_by_name(right_computation_name)[0]  ;
+
+
+        if( left_comp->involved_subset_of_dependences_is_legal(right_comp) == false )
+        {
+                over_all_legality = false;
+                break;
+        }
+
```
Comment (rbaghdadi) [inline src/tiramisu_function.cpp:2658]: Drop space.

File: tests/Testing/Temporary/CTestCostData.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+---
```
Comment (rbaghdadi) [inline tests/Testing/Temporary/CTestCostData.txt:1]: This is an automatically generated file, it should not be included.

File: tests/Testing/Temporary/LastTest.log [hunk 7]
```diff
@@ -0,0 +1,3 @@
+Start testing: Feb 20 13:06 CET
+----------------------------------------------------------
```
Comment (rbaghdadi) [inline tests/Testing/Temporary/LastTest.log:2]: Please remove this file.

File: src/tiramisu_core.cpp (hunk 1 above)
Comment (nassimTchoulak) [reply src/tiramisu_core.cpp:107]: a lot of these methods uses the function object's and it's attributes. But since the function should not be exposed to regular users, I duplicated these. so you will see the function methods where the real execution is done, and a code exposed in tiramisu:: which only gets the implicit function a call the needed method.

Comment (rbaghdadi) [reply src/tiramisu_core.cpp:107]: Yes indeed I saw that wile reading the code.

File: include/tiramisu/core.h [hunk 8]
```diff
@@ -136,6 +136,27 @@ void init();
...
+ * Full check of legality for this function using dependency analysis 
+ * must be used after invoking : performe_full_dependecy_analysis()
+ */
+bool check_legality_of_function();
+
+
+/**
+ * Performe a full dependecy analysis RAW/WAR/WAW and the result is stored in attributes inside the function
+ * To invoke this method user must : define computations order & define the buffer mapped to each computation using :
+ *   tiramisu::prepare_schedules_for_legality_checks()
```
[the remark this answers is not in the document]
Comment (nassimTchoulak) [reply include/tiramisu/core.h:149]: calling tiramisu::prepare_schedules_for_legality_checks() before each legality check (bool check_legality_of_function()) would reduce complexity. there is 2 motivations behind extraction out prepare_schedules_for_legality_checks() :

1. heavy calculation : since prepare_schedules_for_legality_checks() would be required to go through all the computations, and for each one apply a range of make them of the same out dimension. in the regular case this is okey, but what would happen if creating my space would require millions of checks with a high number of computation.
2. small merit : prepare_schedules_for_legality_checks() is only necessary when we apply an optimization that changes the out dimension of a schedule (tiling) or we fuse a computation in a different loop level.

My solution was to let the user manually call tiramisu::prepare_schedules_for_legality_checks() .
another solution could be to call  tiramisu::prepare_schedules_for_legality_checks() only when called from tiramisu::perform_full_dependency_analysis() 
so the part exposed to regular users would be more simple and those using the function object would need the difference.
so what do you think is the best approach ?

[the remark this answers is not in the document]
Comment (rbaghdadi) [reply include/tiramisu/core.h:149]: Ok, let's stick with your 1st solution (i.e., let the user call it). I would rephrase the documentation as follows though (to make the requirements clearer):

Before invoking this method, the user must call tiramisu::prepare_schedules_for_legality_checks() and must define the buffer associated with each computation.
